platform/util/src/com/intellij/util/containers/DisposableWrapperList.java

```diff
@@ -0,0 +1,466 @@
+    return wrapped == null ? -1 : myWrappedList.indexOf(wrapped);
+  }
+
+  @Override
+  public int lastIndexOf(@Nullable Object obj) {
+    DisposableWrapper wrapped = wrap(obj);
+    return wrapped == null ? -1 : myWrappedList.lastIndexOf(wrapped);
+  }
+
+  @Override
+  @NotNull
+  public ListIterator<E> listIterator() {
+    return new DisposableWrapperListIterator<>(myWrappedList, 0);
+  }
+
+  @Override
+  @NotNull
+  public ListIterator<E> listIterator(int index) {
+    return new DisposableWrapperListIterator<>(myWrappedList, index);
+  }
+
+  @Override
+  @NotNull
+  public List<E> subList(int fromIndex, int toIndex) {
+    throw new UnsupportedOperationException();
+  }
+
+  @NotNull
+  private DisposableWrapper<E> createDisposableWrapper(E element, @NotNull Disposable parentDisposable) {
+    DisposableWrapper<E> disposableWrapper = new DisposableWrapper<>(element, myWrappedList);
+    Disposer.register(parentDisposable, disposableWrapper);
+    return disposableWrapper;
+  }
+
+  @Nullable
+  private static <T> DisposableWrapper<T> wrap(@Nullable T obj) {
+    return obj == null ? null : new DisposableWrapper<>(obj);
+  }
+
+  @NotNull
+  private static <T> Collection<DisposableWrapper<T>> wrapAll(@NotNull Collection<? extends T> collection) {
+    if (collection.isEmpty()) {
+      return Collections.emptyList();
+    }
+    List<DisposableWrapper<T>> result = new ArrayList<>(collection.size());
+    for (T obj : collection) {
+      result.add(new DisposableWrapper<>(obj));
+    }
+    return result;
+  }
+
+  @Nullable
+  private E unwrapAndDispose(@Nullable DisposableWrapper<E> disposableWrapper) {
+    if (disposableWrapper == null) {
+      return null;
+    }
+    E unwrapped = disposableWrapper.delegate;
+    disposableWrapper.disposeWithoutRemoval();
+    return unwrapped;
+  }
+
+
+  private static class DisposableWrapper<T> implements Disposable {
+    @NotNull private final T delegate;
+    @Nullable private Collection<DisposableWrapper<T>> container;
+
+    DisposableWrapper(@NotNull T obj) {
+      this(obj, null);
+    }
+
+    DisposableWrapper(@NotNull T obj, @Nullable Collection<DisposableWrapper<T>> container) {
+      this.delegate = obj;
+      this.container = container;
+    }
+
+    @Override
+    public void dispose() {
+      if (container != null) {
+        container.remove(this);
+      }
+    }
+
+    public void disposeWithoutRemoval() {
+      container = null;
+      Disposer.dispose(this);
+    }
+
+    @Override
+    public int hashCode() {
+      return delegate.hashCode();
+    }
+
+    @Override
+    public boolean equals(@Nullable Object obj) {
+      if (obj == this) {
+        return true;
+      }
+      if (obj == null || !getClass().equals(obj.getClass())) {
+        return false;
+      }
+      //noinspection unchecked
+      return delegate.equals(((DisposableWrapper<T>)obj).delegate);
+    }
+  }
+
+  private static class DisposableWrapperListIterator<T> implements ListIterator<T> {
```

Comment:
ConvertingIterator would not work since it doesn't implement the ListIterator interface. I wasn't able to find a suitable ListIterator implementation in the platform code.